lib/puppet/transaction.rb

```diff
@@ -171,6 +171,9 @@ def evaluate(&block)
       report.resources_failed_to_generate = true
     end
 
+    # mark the end of transaction evaluate.
+    report.transaction_completed = true
```

Comment:
Does this need to be explicitly initialized to false sometime earlier? Or would it get initialized properly during instantiation?

Reply:
It should be fine since the default value is `nil` which is "falsey". You're right that it would probably be a good idea to explicitly initialize it to false in `#initialize` to be more intention revealing, though.

This also brings up a good question about how things behave if a report is created via `#initialize_from_hash`. I'm not sure where all that's used, but I noticed it when looking at `#initialize`, and it seems worth looking in to.

Reply:
The `initialize_from_hash` logic is called when deserializing a report, e.g. when read from the network, from disk, etc. If the transaction completion state is supposed to be serialized with the report, then you'll want to update `initialize_from_hash` and `to_data_hash`, update the report schema, and bump the report version (if it hasn't already been bumped for other reasons). If the completion state is supposed to be "volatile", so just used for the duration of the transaction, then it doesn't need to be serialized with the report.